pallets/rewards/src/lib.rs

```diff
@@ -0,0 +1,506 @@
+					if *total_rewards_in_pdex > MIN_REWARDS_CLAIMABLE_AMOUNT {
+						//initialize reward info struct
+						let reward_info = RewardInfoForAccount {
+							total_reward_amount: (*total_rewards_in_pdex).saturated_into(),
+							claim_amount: 0_u128.saturated_into(),
+							is_initial_rewards_claimed: false,
+							is_initialized: false,
+							lock_id: REWARDS_LOCK_ID,
+							last_block_rewards_claim: reward_info.start_block,
+							initial_rewards_claimable: (*initial_rewards_claimable)
+								.saturated_into(),
+							factor: (*factor).saturated_into(),
+						};
+						//insert reward info into storage
+						<Distributor<T>>::insert(reward_id, user.clone(), reward_info);
+					} else {
+						return Err(Error::<T>::AmountToLowtoInitializeRewards.into())
+					}
+				} else {
+					//sanity check
+					return Err(Error::<T>::RewardIdNotRegister.into())
+				}
+			} else {
+				return Err(Error::<T>::UserNotEligible.into())
+			}
+
+			//check if user is added in reward list
+			ensure!(<Distributor<T>>::contains_key(reward_id, &user), Error::<T>::UserNotEligible);
+
+			<Distributor<T>>::mutate(reward_id, user.clone(), |user_reward_info| {
+				if let Some(user_reward_info) = user_reward_info {
+					//check if user already unlocked the rewards
+					//sanity check: if user is calling initialize_for the second time it will fail
+					// in the above ensure, if user is calling initialize for the first time the
+					// flag is set to false by default
+					ensure!(
+						!user_reward_info.is_initialized,
+						Error::<T>::RewardsAlreadyInitialized
+					);
+					//transfer funds from pallet account to users account
+					ensure!(
+						Self::transfer_pdex_rewards(
+							&Self::get_pallet_account(),
+							&user,
+							user_reward_info.total_reward_amount
+						)
+						.is_ok(),
+						Error::<T>::TransferFailed
+					);
+
+					//lock funds in users account
+					T::NativeCurrency::set_lock(
+						REWARDS_LOCK_ID,
+						&user,
+						user_reward_info.total_reward_amount.saturated_into(),
+						WithdrawReasons::TRANSFER,
+					);
+					user_reward_info.is_initialized = true;
+					Ok(())
+				} else {
+					//sanity check
+					Err(Error::<T>::UserNotEligible)
+				}
+			})?;
+			Self::deposit_event(Event::UserUnlockedReward { user, reward_id });
+			Ok(())
+		}
+
+		/// The user will use the extrinsic to claim rewards
+		/// # Parameters
+		/// * `origin`: The users address which has been mapped to reward id
+		/// * `id`: The reward id
+		#[pallet::weight(<T as Config>::WeightInfo::claim())]
+		pub fn claim(origin: OriginFor<T>, reward_id: u32) -> DispatchResult {
+			let user: T::AccountId = ensure_signed(origin)?;
+
+			//check if given id valid or not
+			ensure!(
+				<InitializeRewards<T>>::contains_key(reward_id),
+				Error::<T>::RewardIdNotRegister
+			);
+
+			//check if user is added in reward list
+			ensure!(<Distributor<T>>::contains_key(reward_id, &user), Error::<T>::UserNotEligible);
+
```

Comment:
remove these two checks and return errors instead of "sanity check"s below to reduce reads by 2